WIP: Travis CI Setup Optimizations

jakirkham opened the pull request:
Provides various optimizations to Travis CI's startup. Also cleans up a few things to help make them easier to follow. Highlights of the changes are listed below. More details can be found in the individual commits and their messages.

* Use Homebrew's uninstall script to quietly remove Homebrew in a standard way.
* Fast finish old PR builds when newer ones are queued. 
  * Similar to what is done on AppVeyor ( https://github.com/conda-forge/staged-recipes/pull/123 ).
  * Also inspired by this [script]( https://github.com/JuliaLang/julia/blob/111f5e5a0d18be556060fb4f91a89505f9116ff7/contrib/travis_fastfail.sh ). Thanks @tkelman!
  * Given we do simply this on AppVeyor where we have fewer concurrent builds and it is in better shape, this probably already enough to get Travis CI back to a better place. Though we will need it in the feedstocks too.
* Cache basic conda package tarballs and Miniconda install script to cutdown on startup time.
  * Only deals with basic packages (e.g. what is in `root`).
  * Could extend to allow PRs to cache build related packages as well.

jakirkham commented:
cc @conda-forge/core @goanpeca

jakirkham commented:
FWICT this seems to work ok on Travis CI. However I was running it under Linux on my fork with OS X specific stuff switched to Linux or commented. Had to use Linux on my fork as the Travis CI queue both at conda-forge and for OS X generally is just too full to do any decent iteration. Did try little non-Travis CI specific pieces locally to make sure they worked ok. We'll see how it goes on Travis CI's OS X workers when things simmer down a bit.

jakirkham commented on .CI/travis_fast_finish.py:
```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env python
```
Probably don't want this script going into every feedstock in the event we need to fix it or change it somehow. Maybe it would be better to add it to `conda-forge-build-setup` and then just include the snippet of code to run this script in staged-recipes and the feedstocks.

tkelman replied:
looked better when it was written in shell :-p

jakirkham replied:
Still thinking this should be in `conda-forge-build-setup`. However given this script no longer needs anything beyond a stock Python interpreter, we may just `curl` the script instead of trying to use `conda` to install it. This way we can still run it as early as possible.

tkelman commented:
I wonder where this "Q12017" info is coming from: https://github.com/grosser/travis_dedup/commit/20de01d2e51efed69c450383e75fa66b5b5eca8a

jakirkham commented:
That is very interesting. FWIW I do see a beta feature in the settings panel, but it seems to be unrelated. Has something to do with Cron Jobs, which we might use elsewhere.

jjhelmus commented:
Thanks for putting this together @jakirkham.  Even if this is not a silver bullet for the Travis CI backlog anything that speeds it up is great!

Would it make sense to move the fast finish logic before installing Miniconda to try and save a little more time?  A Python interpreter is likely available in the initial environment, although I'm not sure if requests is installed.

jakirkham commented:
Was thinking the same thing, @jjhelmus. Maybe it could be rewritten to use `urllib` thus avoiding the `requests` dependency. Then I agree we should be able to have it run at the beginning.

jakirkham commented:
Have rewritten this to work with `urllib` and dropped the `requests` dependency as mentioned. Given this change, have now bumped the fast finish step to the top of `build_all` and dropped the rearrangement of the Homebrew step as that is no longer needed. Retested these changes with Travis CI on Linux, which works. Expect it should work fine on OS X, but will need to wait and see.

goanpeca commented on .CI/build_all:
```diff
@@ -1,28 +1,54 @@
+# Add standard packages back to the cache.
+mkdir -p ~/cache/pkgs
+cp -r ~/miniconda/pkgs ~/cache
+
```
Awesome, I did this for a repo the other day and was wondering if we were using it here! :-)

(We can do the same on the other CIs, even if they are not the bottleneck)

jakirkham replied:
Yeah, we used it in some places, but not on Travis CI's OS X builds. Probably because caching was not always supported on OS X.

We could do this on other CIs too. Probably worthwhile on AppVeyor. We switched to their preincluded Miniconda, which helped some. Caching packages should help a little more. CircleCI is already using the Docker image with everything installed so it is effectively cached. I think Docker caching on CircleCI was explored, but was found to be too slow compared to just pulling a fresh image.

goanpeca commented on .CI/create_feedstocks:
```diff
@@ -7,21 +7,32 @@ set -e
+# Reuse packages from the cache.
+mkdir -p ~/cache/pkgs
+rm -rf ~/miniconda/pkgs
+mv ~/cache/pkgs ~/miniconda/pkgs
```
We could try a different approach

Simply define the build cache dir in the conda config so we dont need to move and copy stuff around.

1.) Install conda
2.) Set the package cache to a different place where we set this in the cached

Not sure if conda will complain if the folder already exists :-| ?

jakirkham replied:
Yeah, that could work. We could also just start caching everything like we did [here]( https://github.com/conda-forge/conda-forge.github.io/blob/84f85c4eaa2798b43389d223eac82e3536ea4c75/.travis.yml#L23-L24 ).

After playing with caching a bit, I think we'd be better off to just cache all packages downloaded. PRs have separate caches from `master` and the repo as a whole anyways. So we don't need to worry about getting a massive cache of useless things. Plus it should make it easier for people to iterate on a PR.

goanpeca replied:
Yeah, only packages, the idea was just to avoid the copy/move and make the script simpler

jakirkham replied:
Made some tweaks that should help simplify things a bit. Please let me know what you think.

jakirkham replied:
Another idea might be to make the cache an explicit channel. Not sure how this would effect performance, but it might aid in debugging strange issues.

goanpeca commented:
@jakirkham sorry I havent found time to work on the webservice I started. Is that something that we are still interested in doing? I know the plan is to receive the github push directly and based on some rules, propagate the push notifications to circle, appveyor and travis (supplying some env variables to avoid starting some of the containers.)

I still want to work on that, I just need to find the time :-|

jakirkham commented:
No worries. There's way too much going on here all the time. 😄

Yes, definitely still interested, but no need to feel rushed. Haven't had much time to look at it myself of late either.

Maybe what we can do is try to write up a spec initially given focus has shifted from canceling to filtering. That should hopefully be a less time intensive and more asynchronous activity. Also it should help us get everything organized so that we aren't running into time draining problems at the implementation stage. What do you think?

goanpeca commented:
>Maybe what we can do is try to write up a spec initially given focus has shifted from canceling to filtering. That should hopefully be a less time intensive and more asynchronous activity. Also it should help us get everything organized so that we aren't running into time draining problems at the implementation stage. What do you think?

The way I see it we need both.

1. We need to cancel duplicate builds in a way that is fast and does not need the container to start (multiple PR pushes) that is what the work I did, does and except for the travis rate limit (and some minor issues) it works fine. This work is agnostic as it can be used by anyone and is not conda-forge specific
2. We need to start builds based on the content of the push, so if the recipe only requires PY35, we only launch those containers. This is very specific to conda-forge

jakirkham commented on .CI/create_feedstocks:
```diff
@@ -7,14 +7,21 @@ set -e
-mkdir -p ~/miniconda_staging
-pushd ~/miniconda_staging
-curl -L -O https://raw.githubusercontent.com/conda-forge/conda-smithy/e976c7e84bb3c4846e7562afd1a42c4b535b51f5/bootstrap-obvious-ci-and-miniconda.py
-python bootstrap-obvious-ci-and-miniconda.py ~/miniconda x64 3 --without-obvci && source ~/miniconda/bin/activate root
```
Had to put a hack in to workaround a feedstock conversion issue. So this conflicts, but it is easy to fix.

xref: https://github.com/conda-forge/staged-recipes/pull/2261

jakirkham commented on .CI/create_feedstocks:
```diff
@@ -7,14 +7,21 @@ set -e
+# Install Miniconda.
+echo ""
+echo "Installing a fresh version of Miniconda."
+[ ! -f ~/cache/miniconda.sh ] && curl -L https://repo.continuum.io/miniconda/Miniconda3-latest-MacOSX-x86_64.sh > ~/cache/miniconda.sh
```
This will also need to be Python 2 if we still need the workaround for feedstock conversion.

xref: https://github.com/conda-forge/staged-recipes/pull/2261

jakirkham commented on .CI/build_all:
```diff
@@ -1,28 +1,50 @@
+echo ""
 
 # We just want to build all of the recipes.
 conda-build-all ./recipes --matrix-condition "numpy >=1.11" "python >=2.7,<3|>=3.5"
```
We might need some sort of change here to make sure that packages in the cache are not already used for the build. Alternatively we could add a pre-caching step that clears those packages out of the cache. That all being said, we will need to test if this is required.

jakirkham commented:
Moved the fast finish work into PR ( https://github.com/conda-forge/staged-recipes/pull/2264 ) and generalized it to work on all CIs.

Also am proposing the fast finish script move into `conda-forge-build-setup` in PR ( https://github.com/conda-forge/conda-forge-build-setup-feedstock/pull/52 ). This should make it easier to update and allow near immediate propagation to staged-recipes and feedstocks that use it.

gqmelo commented on .CI/create_feedstocks:
```diff
@@ -7,14 +7,21 @@ set -e
+# Reuse packages from the cache.
+mv -n ~/miniconda/pkgs ~/cache/pkgs
+rm -rf ~/miniconda/pkgs
+ln -s ~/cache/pkgs ~/miniconda/pkgs
```
Wouldn't this cache dependencies downloaded by conda-build too? Or conda-build is downloading packages on a different ```pkgs``` dir?

jakirkham replied:
It should yes. At least that was the hope.

gqmelo replied:
Ok, I forgot that the master branch never runs so every branch will generate its own cache. Then caching all dependencies is not a problem.
But thinking of that, making the master branch run just to generate a miniconda cache is something we could consider. This would make all new branches start with a cache. Anyway, it is just an idea and would be more complex than what is been done here.

jakirkham commented on .travis.yml:
```diff
@@ -34,3 +35,8 @@ script:
+    directories:
+        - $HOME/cache
+before_cache:
+    - find $HOME/cache/pkgs/ -mindepth 1 -maxdepth 1 -type d -exec rm -r {} \;
```
Also @gqmelo this is kind of important. It makes sure we only cache the tarballs of the packages not their decompressed directories.

gqmelo replied:
Thanks, I forgot to handle that. But as I asked on the other PR, wouldn't be better to remove the tarballs instead? This way conda does not need to uncompress them every build

scopatz commented:
This has conflicts. Is it still relevant?

jakirkham commented:
The caching stuff still is. Though I don't know if there is another PR taking that on ATM.

jakirkham commented:
Good call bot! 😄 Closing...
